model/hasfulltablecache.py

```diff
@@ -1,140 +1,91 @@
 # encoding: utf-8
 # HasFullTableCache
 
-import logging
+from collections import namedtuple
+from types import SimpleNamespace
+from typing import Callable, Dict, Hashable, Optional, Tuple
 
-from . import get_one
+from sqlalchemy.orm import Session
 
+from . import Base, get_one
 
-class HasFullTableCache(object):
-    """A mixin class for ORM classes that maintain an in-memory cache of
-    (hopefully) every item in the database table for performance reasons.
-    """
 
-    RESET = object()
+class HasFullTableCache(object):
+    CacheTuple = namedtuple("CacheTuple", ["id", "key", "stats"])
 
-    # You MUST define your own class-specific '_cache' and '_id_cache'
-    # variables, like so:
-    #
-    # _cache = HasFullTableCache.RESET
-    # _id_cache = HasFullTableCache.RESET
+    """
+    A mixin class for ORM classes that maintain an in-memory cache of
+    items previously requested from the database table for performance reasons.
 
-    @classmethod
-    def reset_cache(cls):
-        cls._cache = cls.RESET
-        cls._id_cache = cls.RESET
+    Items in this cache are always maintained in the same database session.
+    """
 
     def cache_key(self):
         raise NotImplementedError()
 
     @classmethod
-    def _cache_insert(cls, obj, cache, id_cache):
-        """Cache an object for later retrieval, possibly by a different
-        database session.
-        """
+    def _cache_insert(cls, obj: Base, cache: CacheTuple):
+        """Cache an object for later retrieval."""
         key = obj.cache_key()
         id = obj.id
-        try:
-            if cache != cls.RESET:
-                cache[key] = obj
-            if id_cache != cls.RESET:
-                id_cache[id] = obj
-        except TypeError as e:
-            # The cache was reset in between the time we checked for a
-            # reset and the time we tried to put an object in the
-            # cache. Stop trying to mess with the cache.
-            pass
-
-    @classmethod
-    def populate_cache(cls, _db):
-        """Populate the in-memory caches from scratch with every single
-        object from the database table.
-        """
-        cache = {}
-        id_cache = {}
-        for obj in _db.query(cls):
-            cls._cache_insert(obj, cache, id_cache)
-        cls._cache = cache
-        cls._id_cache = id_cache
+        cache.id[id] = obj
+        cache.key[key] = obj
 
     @classmethod
-    def _cache_lookup(cls, _db, cache, cache_name, cache_key, lookup_hook):
+    def _cache_lookup(
+        cls,
+        _db: Session,
+        cache: CacheTuple,
+        cache_name: str,
+        cache_key: Hashable,
+        lookup_hook: Callable,
```

Comment:
Maybe we can rename `lookup_hook` and give it a friendlier name like `cache_miss_hook`?

Reply:
Done